src/lustre/lustreTypeChecker.ml

```diff
@@ -0,0 +1,1133 @@
+         | _ -> type_error pos ("Cannot concat non-array type "
+                               ^ string_of_tc_type ty1))
+
+  (* Quantified expressions *)
+  | LA.Quantifier (pos, _, qs, e) ->
+          let extn_ctx = List.fold_left union ctx
+                          (List.map (fun (_, i, ty) -> singleton_ty i ty) qs) in
+          infer_type_expr extn_ctx e 
+
+  (* Clock operators *)
+  | LA.When (_, e, _) -> infer_type_expr ctx e
+  | LA.Current (_, e) -> infer_type_expr ctx e
+  | LA.Condact (pos, c, _, node, args, defaults) ->
+     R.bind (check_type_expr ctx c (Bool pos)) (fun _ ->
+         R.bind (infer_type_expr ctx (Call (pos, node, args))) (fun r_ty ->
+             R.bind (R.seq (List.map (infer_type_expr ctx) defaults)) (fun d_tys -> 
+                 R.bind (eq_lustre_type ctx r_ty (TupleType (pos, d_tys)))(fun is_eq -> 
+                 if is_eq then R.ok r_ty
+                 else type_error pos "Defaults do not have the same type as node call"))))
+  | LA.Activate (pos, node, cond, rcond, args) ->
+     R.bind (check_type_expr ctx cond (Bool pos)) (fun _ ->
+         infer_type_expr ctx (Call (pos, node, args)))
+  | LA.Merge (pos, i, mcases) ->
+     R.bind (infer_type_expr ctx (LA.Ident (pos, i))) (fun _ ->
+         let case_tys = List.map snd mcases |> List.map (infer_type_expr ctx) in
+         R.bind(R.seq case_tys) (fun tys ->
+             let main_ty = List.hd tys in
+             R.bind(R.seq (List.map (eq_lustre_type ctx main_ty) tys)) (fun isEqs -> 
+                 if
+                   List.fold_left (&&) true isEqs
+                 then R.ok main_ty
+                 else type_error pos ("All expressions in merge expected to be the same type "
+                                     ^ string_of_tc_type main_ty))))
+  | LA.RestartEvery (pos, node, args, cond) ->
+     R.bind(check_type_expr ctx cond (LA.Bool pos)) (fun _ ->
+         infer_type_expr ctx (LA.Call (pos, node, args)))
+                                
+  (* Temporal operators *)
+  | LA.Pre (pos, e) -> infer_type_expr ctx e
+  | LA.Last (pos, i) -> infer_type_expr ctx (LA.Ident (pos, i))
+  | LA.Fby (pos, e1, _, e2) ->
+     R.bind (infer_type_expr ctx e1) (fun ty1 ->
+         R.bind(infer_type_expr ctx e2)(fun ty2 ->
+             R.bind(eq_lustre_type ctx ty1 ty2)(fun is_eq ->
+                 if is_eq
+                 then R.ok ty1
+                 else type_error pos ("Both the expressions in Fby should be of the same type."
+                                     ^ "Found types " ^ string_of_tc_type ty1
+                                     ^ " and " ^ string_of_tc_type ty2))))
+  | LA.Arrow (pos, e1, e2) ->
+     R.bind (infer_type_expr ctx e1) (fun ty1 ->
+         R.bind (infer_type_expr ctx e2) (fun ty2 ->
+             R.bind((eq_lustre_type ctx ty1 ty2)) (fun is_eq -> 
+                 if is_eq then R.ok ty1 else
+                   type_error pos
+                     ("Arrow types do not match "
```

Comment:
Test type checker over `tests/regression/success/abstract_type.lus`. It returns:

> line 22 col. 6: Failure cause: Type error: Arrow types do not match COUNTER and [COUNTER]

Expected output: 

> No type errors found!

Reply:
This is fixed now.